### ISSUE TITLE
Allow data loading in REST integrations

### ISSUE THREAD

**Issue by Mobe91** [Blazebit/blaze-persistence#1213]
> There are use cases where we would like to actually load the updatable view and deserialize onto that loaded instance.
> We could tell users to inject the EntityViewAwareObjectMapper if that is possible, and use it like this
> ```
> @Inject EntityViewAwareObjectMapper mapper;
> 
> @Path("{id}")
> public Response put(@PathParam("id") Long id, InputStream is) {
>   MyView v = repository.findById(id);
>   mapper.readerFor(MyView.class).withValueToUpdate(v);
> }
> ```
> That is IMO fine if it's possible, but should definitely be documented.

Quote @beikov

**Comment by Mobe91** [comment]
What do you think about:
```java
@Path("{id}")
public Response put(@Fetch @EntityViewId("id") MyView v) {
}
```

**Comment by beikov** [comment]
I think I'd like an annotation like e.g. `@EntityViewResource(idParamName = "id", applyMode = ApplyMode.AUTO)` with e.g. the following modes:

* `AUTO` - models what we have now
* `CREATE` - always use create
* `PATCH` - apply on the reference
* `UPDATE` - apply on the loaded object

**Comment by Mobe91** [comment]
@beikov 
Where would this annotation be placed? Can you provide a complete example? Would this mean that the @EntityViewId annotation becomes obsolete?

In essence, what is the functional difference between your suggested `PATH` and `UPDATE` modes and what is the reasoning behind their names?

**Comment by beikov** [comment]
On the entity view parameter, so yes, the @EntityViewId annotation would become obsolete.

UPDATE would apply the json payload on a view loaded via `entityViewManager.find`.
This allows to replace collections or remove elements from collections whereas PATCH only allows adding elements currently

**Comment by Mobe91** [comment]
I don't get it.
Consider an endpoint PATCH /persons/{id} that accepts JSON representing a PersonView. `contacts` is an updatable collection mapping within PersonView.
```json
{
  "id": 12,
  "contacts": [
    { "id": 1, "name": "A"},
    { "id": 2, "name": "B"}
  ]
}
```
I would expect the persisted contacts collection to be _replaced_ by the contacts collection in above request. There is not even a way to denote that I want these elements to be removed/added to the persisted collection rather than replacing the whole collection. I think that's why I don't get the point of your proposal.

**Comment by beikov** [comment]
You are right, the JSON mapping does `clear` + `add` to the collection, so yes, it would replace the collection when using UPDATE.
The difference is, that PATCH which uses a reference, would allow to only add elements but then it also doesn't load the collection.

**Comment by Mobe91** [comment]
So since the current implementation uses `getReference`, applying the above JSON leads to the contacts being added to any existing contacts? That's fucked up. 😂

**Comment by Mobe91** [comment]
IMHO a PATCH mode would need to also support the removal of elements. I don't see how it is useful otherwise.

**Comment by beikov** [comment]
>So since the current implementation uses `getReference`, applying the above JSON leads to the contacts being added to any existing contacts?

You can use `saveFull` to do this replacing even with the `getReference` approach.

>IMHO a PATCH mode would need to also support the removal of elements. I don't see how it is useful otherwise.

I think it's useful for at least updating basic attributes, but maybe this mode should not be the default configuration.

**Comment by beikov** [comment]
I just tried out to add an element to a collection of a reference entity view and it seems I was wrong. The semantic is always the same i.e. replace but the difference with reference views is that only what is dirty is updated *and* it avoids loading data.

**Comment by Mobe91** [comment]
So if the semantics are always the same, there is no need for the ´UPDATE´ mode, right? In that case we can close this issue.

**Comment by beikov** [comment]
Although the flush semantics might be the same, there might be logic in setters or in the service that takes the view for flushing that depends on other state. So the UPDATE mode does make sense for these cases.

**Comment by Mobe91** [comment]
Ok.

I suggest `@EntityViewResource(idParamName = "id", applyMode = ApplyMode.AUTO)` with the following modes:

* `AUTO` - models what we have now
* `CREATE` - always use create
* `UPDATE_REFERENCE` - apply on the reference
* `UPDATE_LOADED` - apply on the loaded object

**Comment by beikov** [comment]
Sounds good! Do you think it makes sense to still use `EntityViewManager.create` if no id is given when the view is updatable and creatable and `UPDATE_REFERENCE` or `UPDATE_LOADED` are used? What I'm trying to ask is, should the configuration force either create or update or should it just provide a preference for how to update? I think for POST endpoints it might make sense to allow creation as well. If creating new objects is not desired, the view should simply not be creatable.

**Comment by Mobe91** [comment]
Ok, let me revise my suggestion:

```
+-------------------------+----------------+-------------------+----------------+-------------------+-----------------------------+--------------------------------+
|                         | creatable (id) | creatable (no id) | updatable (id) | updatable (no id) | creatable + updateable (id) | creatable + updateable (no id) |
+-------------------------+----------------+-------------------+----------------+-------------------+-----------------------------+--------------------------------+
| AUTO                    | create         | ERROR             | update ref     | ERROR             | update ref                  | create                         |
| CREATE                  | create         | ERROR             | ERROR          | ERROR             | create                      | create                         |
| UPDATE_REFERENCE        | ERROR          | ERROR             | update ref     | ERROR             | update ref                  | create                         |
| UPDATE_LOADED           | ERROR          | ERROR             | update loaded  | ERROR             | update loaded               | create                         |
| STRICT_UPDATE_REFERENCE | ERROR          | ERROR             | update ref     | ERROR             | update ref                  | ERROR                          |
| STRICT_UPDATE_LOADED    | ERROR          | ERROR             | update loaded  | ERROR             | update loaded               | ERROR                          |
+-------------------------+----------------+-------------------+----------------+-------------------+-----------------------------+--------------------------------+
```
This should cover all use cases. Do you agree?

**Comment by beikov** [comment]
I suggest two settings, `ApplyMode` and a boolean flag `updateReference`.

```
+--------------------------------+----------------+-------------------+----------------+-------------------+-----------------------------+--------------------------------+
|                                | creatable (id) | creatable (no id) | updatable (id) | updatable (no id) | creatable + updateable (id) | creatable + updateable (no id) |
+--------------------------------+----------------+-------------------+----------------+-------------------+-----------------------------+--------------------------------+
| CREATE                         | create         | create            | ERROR          | ERROR             | create                      | create                         |
| AUTO + updateReference=true    | create         | create            | update ref     | ERROR             | update ref                  | create                         |
| AUTO + updateReference=false   | create         | create            | update loaded  | ERROR             | update loaded               | create                         |
| UPDATE + updateReference=true  | ERROR          | ERROR             | update ref     | ERROR             | update ref                  | ERROR                          |
| UPDATE + updateReference=false | ERROR          | ERROR             | update loaded  | ERROR             | update loaded               | ERROR                          |
+--------------------------------+----------------+-------------------+----------------+-------------------+-----------------------------+--------------------------------+
```
